doc/designs/migration.md

```diff
@@ -0,0 +1,456 @@
+is maintained then it can be easily calculated which are ready for migration and
+which are not without having to refer to LDAP.
+
+#### Objects to ignore
+
+The IPA API object list includes an number of internal-only objects and
+some classes that represent informational commands so they can be skipped
+for migration.
+
+| Object |	Reason |
+| --- | ----- |
+| certreq | no storage |
+| class | schema |
+| command | internal |
+| cosentry | migration not needed |
+| dns_system_records | no storage |
+| metaobject | schema |
+| param | internal |
+| pkinit | no storage |
+| topic | internal |
+| userstatus | no storage |
+
+#### Objects TBD
+
+Each DNS record type is represented as a separate class. It is probably safe
+to just bulk import all dns records as there isn't any IPA-specific information in them.
+So basically add them to the ignore list.
+
+#### Other Containers
+
+Not all data in IPA can be manipulated with the API.
+
+cn=certificates,cn=ipa,cn=etc,$SUFFIX contains any user-provided certificates.
+I think we can skip this and require that any necessary certificates be
+re-added. Otherwise, to be safe, we should validatate the trust and expiration
+dates on all of them.
+
+DNA ranges. Depending on the local ranges we could try to recover them.
+
+#### Items to NOT be migrated
+
+We will not have migrate existing keys from the remote IPA server so the
+following will not be migrated:
+
+* Self-signed CA certificates (maybe we import user-added CA certs)
+* KRA keys
+* sub-CAs
+* caacls
+* DNSSEC keys
+* admin password
+* DM password
+* Basically anything in Custodia
+
+#### Migrating conflicting data
+
+For new entries (users, groups, hosts, etc) we can do a straight LDAP
+ADD after massaging necessary attributes (see Attributes to not migrate).
+
+That is the easy part.
+
+There are some parts of IPA that are default and will exist in both servers
+such as the configuration (config commands), allow_all HBAC rule,
+permissions, privileges, roles, selinux mappings, ipaservers hostgroup and more.
+
+The strategy for this will be use the existing entry comparison in ldap2 and
+modify the new IPA instation to match the original minus some specific things like
+the CA subject base in the config. So in most cases we favor the remote IPA
+installation except for things which are installation-dependent.
+
+The downside of this is that it could interpret "fresher" defaults in the
+new IPA server with stale ones from the old. We will need good logging around
+these types of entries.
+
+Standard boilerplate will not be evaluated, things like cosTemplates. These are not
+user-modifiable and if a user decides to mess with them, they can correct it
+after migration.
+
+#### Migrating cn=config
+
+A number of plugins are optional in IPA: nis, compat, ACME.
+
+These will be enabled in the new IPA server if enabled in the remote:
+
+* cn=Schema Compatibility, cn=plugins, cn=config
```

Comment:
Note: if the schema compat plugin is enabled, entries below cn=compat,$BASEDN must be skipped by the migration command. They would be dynamically re-created on the target LDAP server.